src/sources/socket/udp.rs

```diff
@@ -53,6 +65,31 @@ pub fn udp(
             .expect("Failed to bind to udp listener socket");
         info!(message = "Listening.", address = %address);
 
+        {
+            // SAFETY: We temporarily take ownership of the socket and return it by the end of this block scope.
+            let socket = unsafe { socket2::Socket::from_raw_fd(socket.as_raw_fd()) };
```

Comment:
This will be obsolete once tokio/mio use `socket2` and propagate the corresponding methods.

Reply:
Are we actually taking ownership of the original `socket` here? I worry that if `as_raw_fd()` is just borrowing then we could end up with a double free (double close?) when both of these are dropped.

Reply:
This looks weird, but I think it's ok. The `socket.into_raw_fd();` at the end prevents dropping `socket` from closing the FD. Some comment there would help clear that up. This could also be cleared up by *actually* taking ownership with `into_raw_fd` and re-emitting the socket with `into_udp_socket()`, but if this section will go away later it's not worth engineering too carefully.

It would be good to add a comment about this being obsoleted once tokio exports support for this, though.

Reply:
It's a valid concern, thanks for looking at it carefully. It's doing what @bruceg said, I'm updating the corresponding comment.

I don't like this either, but I don't want to rearchitect the part to destroy and recreate the actual sockets, since that will not be needed after the upgrade.